### ISSUE TITLE
Support setting headers per request with parameters from map 

### ISSUE THREAD

**Issue by nmiyake** [Netflix/feign#363]
Analogous to #330 , but for headers instead of query parameters.

Proposal is to introduce a `@HeaderMap` parameter annotation. The annotation causes the contents of the map to be used as the header parameters for the request. The contents of the map will be additive to header parameters specified by other annotations.

For example:

```
@RequestLine("POST /")
@Headers({"Auth-Token: {authToken}", "Auth-Token: Foo"})
void logout(@Param("authToken") String token, @HeaderMap Map<String, Object> headers);
```

In this example, any calls will expand the content of `headers` and add it (or override if keys are identical) to the headers for the call.

**Comment by nmiyake** [comment]
Although the approach is slightly different, this feature may also address issues raised in #214 and #288

**Comment by nmiyake** [comment]
Actually, on further reading realized that this won't address #214 since in that case there's a desire to explicitly not have header parameters as any part of the method parameters.

Also, definitely open to discussion/other approaches for accomplishing this. However, much like the `QueryParams` annotation, in a Feign ecosystem I think this approach is fairly clean and provides the ability to handle headers more dynamically at the API level instead of having to write targets/interceptors that are conceptually more client-specific instead of API-specific

**Comment by nmiyake** [comment]
@adriancole for feedback

**Comment by adriancole** [comment]
The problem with HeaderMap, is that it continues the slippery slope away
from templated requests and purpose-made interceptors or targets. I foresaw
a bit of this with QueryMap. Now, there are at least 4 ways to add a header
value dynamically.

I think this type should be explicitly documented for the edge case that it
solves, which is that you have a request-scoped header, and you don't have
anything like a thread local context to attach it with. If it isn't
edge-case only, a concrete example of how interceptor or target doesn't
work would help. Ex real life apis.

So, I'd update both the PR description and the javadoc on HeaderMap with
compelling reasons why we are adding this, vs deferring folks to look at
other issues.

Not only would this reduce confusion when one wonders which of the several
ways they should use, but also give sufficient background on why we
continue this deviation from api minimalism.

**Comment by nmiyake** [comment]
Thanks for the feedback -- I will look at the other options and think through this a bit more. Will then either update PR with description/Javadoc with better explanation of the case this addresses (and why it is needed despite the other available mechanisms) or close PR.

**Comment by adriancole** [comment]
in hind sight, I think we at least have a documentation gap. Maybe a section explaining the few ways to set a header could help regardless of where this goes.

**Comment by nmiyake** [comment]
Thought about this a bit more, and opened #366 with documentation for current approaches (not sure if README is the best place, but that seems to be where most of current documentation is located -- let me know if there's a better place for this).

After writing that up, I think there is a place for `@HeaderMap`. It basically falls in category 1 (setting headers as part of the API), but adds a dynamic element to it. Although it may be a corner case, it provides some distinct things not offered by `RequestInterceptor` or `Target` -- it makes specifying headers part of the core API, which offers the following benefits:
* Setting headers is thread-safe
* Setting headers can be done cleanly on a per-request basis in a thread-safe manner
* Setting headers is part of the API rather than the client, so existing client setup code (which may already have their own custom interceptors or targets) can use it seamlessly without having to add more interceptors or targets

I'll update this PR with better documentation to explain this motivation/use case. If #366 merges, I will modify the "Headers" section there to include this usage as well.

**Comment by adriancole** [comment]
Thanks for adding rationale. One key thing is that this feature is only
needed for unknown headers, right? Known header keys can be set thread safe
etc at the moment.

So in this case, headers unknown at the time of designing the API can be
added without using things like thread local interceptors.

Mind putting some examples where an unknown header is part of the API? It
is more common in case of query parameters.

I can think of one, for example user metadata encoded as headers. Ex Amazon
S3 objects, if I'm not mistaken. You wouldn't know the range of keys up
front.

I think a solid use case like this will drive home where it is needed this,
while hopefully still encouraging templates when header keys are known.

Thanks for thinking this through.. I'll carve time tomorrow to review
everything.

**Comment by nmiyake** [comment]
Sure, will update comments. As you mentioned, metadata is an existing case for this, and exists for both AWS (`x-amz-meta-*`, http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectHEAD.html) and Google Cloud (`x-goog-meta-*`, https://cloud.google.com/storage/docs/reference-headers#xgoogmeta).

Another use case is conditional headers. For example, for server-side encryption requests in AWS (http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html), if the value of `x-amz-server-side​-encryption` is `aws:kms`, then `x-amz-server-side-encryption-aws-kms-key-id` must be set as a header value as well. This kind of conditional/dependency based on header field value is hard to express in a declarative manner even if the header field keys themselves are known.

**Comment by adriancole** [comment]
Love the examples.

**Comment by nmiyake** [comment]
Updated PR based on docs commit that merged earlier. Let me know if there's anything else I can do for this on my end (including discussing further if necessary :)

**Comment by adriancole** [comment]
all good. nice work!

**Comment by nmiyake** [comment]
Fixed by #362
